In [2]:
#importing necessary libraries

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns 

%matplotlib inline 

In [3]:
#reading csv file 
ev = pd.read_csv("data/Electric_Vehicle_Population_Data.csv")

display(ev.head())
display(ev.info())

VIN (1-10)    County     City State  Postal Code  Model Year    Make  \
0  5YJXCAE26J    Yakima   Yakima    WA      98908.0        2018   TESLA   
1  JHMZC5F37M    Kitsap  Poulsbo    WA      98370.0        2021   HONDA   
2  5YJ3E1EB0K      King  Seattle    WA      98199.0        2019   TESLA   
3  1N4AZ0CP5D      King  Seattle    WA      98119.0        2013  NISSAN   
4  5YJSA1E21H  Thurston    Lacey    WA      98516.0        2017   TESLA   

     Model                   Electric Vehicle Type  \
0  MODEL X          Battery Electric Vehicle (BEV)   
1  CLARITY  Plug-in Hybrid Electric Vehicle (PHEV)   
2  MODEL 3          Battery Electric Vehicle (BEV)   
3     LEAF          Battery Electric Vehicle (BEV)   
4  MODEL S          Battery Electric Vehicle (BEV)   

  Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
0           Clean Alternative Fuel Vehicle Eligible             238   
1           Clean Alternative Fuel Vehicle Eligible              47   
2           Clean Alternative Fuel Vehicle Eligible             220   
3           Clean Alternative Fuel Vehicle Eligible              75   
4           Clean Alternative Fuel Vehicle Eligible             210   

   Base MSRP  Legislative District  DOL Vehicle ID  \
0          0                  14.0       141151601   
1          0                  23.0       171566447   
2          0                  36.0         9426525   
3          0                  36.0       211807760   
4          0                  22.0       185810306   

              Vehicle Location                              Electric Utility  \
0  POINT (-120.56916 46.58514)                                    PACIFICORP   
1  POINT (-122.64681 47.73689)                        PUGET SOUND ENERGY INC   
2  POINT (-122.40092 47.65908)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
3   POINT (-122.3684 47.64586)  CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)   
4  POINT (-122.75379 47.06316)                        PUGET SOUND ENERGY INC   

   2020 Census Tract  
0       5.307700e+10  
1       5.303509e+10  
2       5.303301e+10  
3       5.303301e+10  
4       5.306701e+10

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130443 entries, 0 to 130442
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         130443 non-null  object 
 1   County                                             130440 non-null  object 
 2   City                                               130440 non-null  object 
 3   State                                              130443 non-null  object 
 4   Postal Code                                        130440 non-null  float64
 5   Model Year                                         130443 non-null  int64  
 6   Make                                               130443 non-null  object 
 7   Model                                              130221 non-null  object 
 8   Electric Vehicle Type                              130443 non-null  object

None

In [5]:
#checking for null values 
ev.isna().sum()

VIN (1-10)                                             0
County                                                 3
City                                                   3
State                                                  0
Postal Code                                            3
Model Year                                             0
Make                                                   0
Model                                                222
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         0
Base MSRP                                              0
Legislative District                                 305
DOL Vehicle ID                                         0
Vehicle Location                                      33
Electric Utility                                       3
2020 Census Tract                                      3
dtype: int64

The following features have missing values:
1. `County` 
2. `City`
3. `Postal Code`
4. `Model`
5. `Legislative District`
6. `Vehicle Location`
7. `Electric Utility`
8. `2020 Census Tract`

with Model and Legislative District having the most missing values with 222 and 305 respectively. 

In [8]:
#taking a closer look at Legislative District
len(ev['Legislative District'].value_counts())

49

In [18]:
#taking a look at the instances that has null values for the Legislative District
null_ld = ev[ev['Legislative District'].isna()]
null_ld.head()

VIN (1-10)      County        City State  Postal Code  Model Year  \
102    1N4AZ0CP1D         NaN         NaN    AP          NaN        2013   
103    5YJ3E1EA5K         NaN         NaN    BC          NaN        2019   
57726  2C4RC1H78J      Oldham      Goshen    KY      40026.0        2018   
59729  WA1VABGE7M     Douglas       Omaha    NE      68102.0        2021   
59939  1N4AZ1CP4J  Montgomery  Norristown    PA      19403.0        2018   

           Make     Model                   Electric Vehicle Type  \
102      NISSAN      LEAF          Battery Electric Vehicle (BEV)   
103       TESLA   MODEL 3          Battery Electric Vehicle (BEV)   
57726  CHRYSLER  PACIFICA  Plug-in Hybrid Electric Vehicle (PHEV)   
59729      AUDI    E-TRON          Battery Electric Vehicle (BEV)   
59939    NISSAN      LEAF          Battery Electric Vehicle (BEV)   

      Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
102             Clean Alternative Fuel Vehicle Eligible              75   
103             Clean Alternative Fuel Vehicle Eligible             220   
57726           Clean Alternative Fuel Vehicle Eligible              33   
59729           Clean Alternative Fuel Vehicle Eligible             222   
59939           Clean Alternative Fuel Vehicle Eligible             151   

       Base MSRP  Legislative District  DOL Vehicle ID  \
102            0                   NaN       112277063   
103            0                   NaN       475254825   
57726          0                   NaN       277234681   
59729          0                   NaN       137619352   
59939          0                   NaN       349481695   

                                   Vehicle Location  \
102                                             NaN   
103                                             NaN   
57726                    POINT (-85.57785 38.40025)   
59729                    POINT (-95.94026 41.26068)   
59939  POINT (-75.39504169999998 40.14662990000005)   

                            Electric Utility  2020 Census Tract  
102                                      NaN                NaN  
103                                      NaN                NaN  
57726  NON WASHINGTON STATE ELECTRIC UTILITY       2.118503e+10  
59729  NON WASHINGTON STATE ELECTRIC UTILITY       3.105500e+10  
59939  NON WASHINGTON STATE ELECTRIC UTILITY       4.209120e+10

In [19]:
#state from null_ld 
null_ld['State'].value_counts()

CA    83
VA    35
MD    29
TX    19
CO    11
NC    10
AZ     9
IL     8
CT     7
NV     7
NY     6
OR     6
GA     6
NJ     5
SC     5
HI     5
LA     4
DC     4
FL     4
NE     3
MA     3
KS     3
AR     3
PA     2
UT     2
ID     2
WY     2
TN     2
BC     2
MO     2
OH     2
OK     1
IN     1
MS     1
NH     1
WI     1
DE     1
NM     1
RI     1
AK     1
MN     1
KY     1
MI     1
AL     1
AP     1
Name: State, dtype: int64

I was right! These null values from Legislative District are all states that are not Washington. Since this is only focusing on the state of WA right now, I am going to drop these `Legislative District` null values.

In [21]:
#dropping null values
ev.dropna(subset=['Legislative District'], inplace=True)

#checking null values 
ev.isna().sum()

VIN (1-10)                                             0
County                                                 0
City                                                   0
State                                                  0
Postal Code                                            0
Model Year                                             0
Make                                                   0
Model                                                222
Electric Vehicle Type                                  0
Clean Alternative Fuel Vehicle (CAFV) Eligibility      0
Electric Range                                         0
Base MSRP                                              0
Legislative District                                   0
DOL Vehicle ID                                         0
Vehicle Location                                      27
Electric Utility                                       0
2020 Census Tract                                      0
dtype: int64

Dropping the null values from Legislative District took away other null values expect for all of `Make` and some `Vehicle Location`

In [26]:
#taking closer look at instances that have Model as null
null_m = ev[ev['Model'].isna()]

null_m.head()

VIN (1-10) County         City State  Postal Code  Model Year   Make  \
533   YV4ED3UL9P   King  Woodinville    WA      98072.0        2023  VOLVO   
1740  YV4ED3UW2P   King     Bellevue    WA      98005.0        2023  VOLVO   
2158  YV4ED3UM9P   King      Seattle    WA      98122.0        2023  VOLVO   
2221  YV4ED3UW1P   King      Seattle    WA      98125.0        2023  VOLVO   
2959  YV4ED3UW1P   King      Seattle    WA      98146.0        2023  VOLVO   

     Model           Electric Vehicle Type  \
533    NaN  Battery Electric Vehicle (BEV)   
1740   NaN  Battery Electric Vehicle (BEV)   
2158   NaN  Battery Electric Vehicle (BEV)   
2221   NaN  Battery Electric Vehicle (BEV)   
2959   NaN  Battery Electric Vehicle (BEV)   

      Clean Alternative Fuel Vehicle (CAFV) Eligibility  Electric Range  \
533   Eligibility unknown as battery range has not b...               0   
1740  Eligibility unknown as battery range has not b...               0   
2158  Eligibility unknown as battery range has not b...               0   
2221  Eligibility unknown as battery range has not b...               0   
2959  Eligibility unknown as battery range has not b...               0   

      Base MSRP  Legislative District  DOL Vehicle ID  \
533           0                  45.0       233585838   
1740          0                  41.0       224709985   
2158          0                  43.0       227352583   
2221          0                  46.0       236425430   
2959          0                  34.0       227240163   

                 Vehicle Location  \
533   POINT (-122.15545 47.75448)   
1740   POINT (-122.1621 47.64441)   
2158  POINT (-122.31009 47.60803)   
2221   POINT (-122.3026 47.72656)   
2959  POINT (-122.36178 47.49408)   

                                   Electric Utility  2020 Census Tract  
533   PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
1740  PUGET SOUND ENERGY INC||CITY OF TACOMA - (WA)       5.303302e+10  
2158   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10  
2221   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303300e+10  
2959   CITY OF SEATTLE - (WA)|CITY OF TACOMA - (WA)       5.303301e+10

Taking a closer look there's seems to be some connection between the make being Volvo and the model being null. As well as the electric vehicle type and the clean alternative fuel vehicle.

In [31]:
display(null_m['Make'].value_counts())

display(null_m['Electric Vehicle Type'].value_counts())

display(null_m['Clean Alternative Fuel Vehicle (CAFV) Eligibility'].value_counts())

display(null_m['Model Year'].value_counts())

VOLVO    222
Name: Make, dtype: int64

Battery Electric Vehicle (BEV)    222
Name: Electric Vehicle Type, dtype: int64

Eligibility unknown as battery range has not been researched    222
Name: Clean Alternative Fuel Vehicle (CAFV) Eligibility, dtype: int64

2023    222
Name: Model Year, dtype: int64

From here can see that all of the null values have the following as Make - Volvo, Electric Vehicle Type - Battery Electric Vehicle (BEV), Clean Alternative Fuel Vehicle (CAFV) Eligibility, Year - 2023. 

Could be that since the cars are new cars (at the time of this analysis) then that is why the model is not know yet. I will not remove this since it gives nice insights into the data.

In [33]:
#dropping null values of Vehicle Location since there is only 27 missing 
ev.dropna(subset=['Vehicle Location'], inplace=True)

In [34]:
ev.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 130111 entries, 0 to 130442
Data columns (total 17 columns):
 #   Column                                             Non-Null Count   Dtype  
---  ------                                             --------------   -----  
 0   VIN (1-10)                                         130111 non-null  object 
 1   County                                             130111 non-null  object 
 2   City                                               130111 non-null  object 
 3   State                                              130111 non-null  object 
 4   Postal Code                                        130111 non-null  float64
 5   Model Year                                         130111 non-null  int64  
 6   Make                                               130111 non-null  object 
 7   Model                                              129889 non-null  object 
 8   Electric Vehicle Type                              130111 non-null  object

Now instead of 130,443 entries have 130,111 which is not bad. 